src/wp-includes/html-api/class-wp-html.php

```diff
@@ -0,0 +1,19 @@
+<?php
+
+class WP_HTML {
+	public static function make_tag( $tag_name, $attributes = null, $data = '' ) {
```

Comment:
In a sperate PR I'd love to add `make_class_name` that would do what `classnames()` does in React.

Reply:
```php
class WP_CSS {
	public static function make_class( $class_names ) {
		
	}
}
```

this is another reason it could be useful to expose the processor rather than returning a string. the need to call `->get_updated_html()` on it is the downside.

```php
$panel = WP_HTML::make_tag( 'div', array( 'class' => 'wp-block-group' ) );
if ( $is_fullscreen ) {
	$panel->add_class( 'fullscreen' );
}
```

Reply:
The alternatives:

```php
WP_HTML::make_tag(
    'div',
    WP_HTML::make_class( 'wp-block-group', [ 'fullscreen' => $is_fullscreen ] )
);
```

Or:

```php
$panel = new WP_HTML_Tag_Processor(
    WP_HTML::make_tag( 'div' )
);
if ( $is_fullscreen ) {
	$panel->add_class( 'fullscreen' );
}
```

> the need to call ->get_updated_html() on it is the downside.

Not at all, `__toString()` is still there!

Reply:
Now, let's consider conditional attributes:



### Attributes array

```php
$attributes = [
    'src' => $src
    'alt' => $has_alt ? $alt : false,
];
if ( ! $only_aria ) {
    $attributes['aria-title'] = $title;
} else {
    $attributes['title'] = $title;
}
WP_HTML::make_tag( 'div', $attributes );
```

### Returning a Tag Processor

```php
$panel = WP_HTML::make_tag( 'div', array( 'src' => $src ) );
if ( $has_alt ) {
	$panel->set_attribute( 'alt', $alt );
}
if ( $only_aria ) {
	$panel->set_attribute( 'aria-title', $title );
} else {
	$panel->set_attribute( 'title', $title );
}
```

### Manual wrapping in a tag processor

```php
$panel = new WP_HTML_Tag_Processor(
    WP_HTML::make_tag( 'div', array( 'src' => $src ) )
);
// ... The same as above
```

Reply:
What stands out:

* The array approach is more concise and seems to support all possible use-cases
* The tag processor approach still accepts an array of attributes but then requires using another API to handle conditionals

Reply:
Here's a wild idea:


```php

$b = new WP_HTML_Builder();
$b->add_tag( 'form', [ 'class' => 'main' ] )
    ->add_child( 'input', [ 'type' => 'text', 'placeholder' => 'Your query' ] ) 
    ->add_sibling( 'button', [ 'class' => 'primary' ], 'Submit' )
    ->parent()
    ->add_sibling( 'img', [ 'src' => 'logo.png' ] )
;
echo $b;
/*
<form class="main">
    <input type="text" placeholder="Your query" />
    <button class="primary">Submit</button>
</form>
<img src="logo.png" />
*/

```

Reply:
This would play nicely with the future HTML Processor:

```php
$p = new WP_HTML_Processor('<h1 title="main">Site title</h1>');
$p->next_tag();
$p->insert_html_after(
    WP_HTML_Builder::create()
      ->add_tag( 'form', [ 'class' => 'main' ] )
      ->add_sibling( 'img', [ 'src' => 'logo.png' ] )
);
```